### PR TITLE
Add `exceededTimeoutLimit` 3DS error

### PR DESCRIPTION
**Pull request by stechiu**
### Summary of changes

- Added `exceededTimeoutLimit` error code and error message

### Checklist

- [x] Added a changelog entry

- @stechiu

### CODE REVIEW

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift [hunk 1]
```diff
@@ -122,14 +122,22 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
...
                 errorCode = BTThreeDSecureError.failedAuthentication("").errorCode
+                completionHandler(nil, BTThreeDSecureError.failedAuthentication(""))
+            }
+            else if validateResponse.actionCode == .timeout {
+                errorCode = BTThreeDSecureError.exceededTimeoutLimit.errorCode
+                userInfo = [NSLocalizedDescriptionKey: BTThreeDSecureError.exceededTimeoutLimit.localizedDescription]
+                completionHandler(nil, BTThreeDSecureError.exceededTimeoutLimit)
+            }
+            else {
+                completionHandler(nil, NSError(domain: BTThreeDSecureError.errorDomain, code: errorCode, userInfo: userInfo))
```
Comment (scannillo) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:138]: We can follow the simpler completionHandler syntax on line 130
```suggestion
                completionHandler(nil, BTThreeDSecureError.exceededTimeoutLimit)
```

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift [hunk 2]
```diff
@@ -122,14 +122,22 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
...
             if validateResponse.errorNumber == 1050 {
                 errorCode = BTThreeDSecureError.failedAuthentication("").errorCode
+                completionHandler(nil, BTThreeDSecureError.failedAuthentication(""))
+            }
+            else if validateResponse.actionCode == .timeout {
+                errorCode = BTThreeDSecureError.exceededTimeoutLimit.errorCode
+                userInfo = [NSLocalizedDescriptionKey: BTThreeDSecureError.exceededTimeoutLimit.localizedDescription]
+                completionHandler(nil, BTThreeDSecureError.exceededTimeoutLimit)
+            }
+            else {
```
Comment (scannillo) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:137]:
```suggestion
            } else if validateResponse.actionCode == .timeout {
                errorCode = BTThreeDSecureError.exceededTimeoutLimit.errorCode
                userInfo = [NSLocalizedDescriptionKey: BTThreeDSecureError.exceededTimeoutLimit.localizedDescription]
                completionHandler(nil, BTThreeDSecureError.exceededTimeoutLimit)
            } else {
```

See the `Spacing` section in our internal team style guide (in Slack)

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift (hunk 1 above)
Comment (stechiu) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:138]: thanks!

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift (hunk 2 above)
Comment (stechiu) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:137]: Noted, thanks!

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift [hunk 3]
```diff
@@ -122,14 +122,20 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
                 completion: completionHandler
             )
         case .error, .timeout:
-            let userInfo = [NSLocalizedDescriptionKey: validateResponse.errorDescription]
+            var userInfo = [NSLocalizedDescriptionKey: validateResponse.errorDescription]
```
Comment (richherrera) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:125]: Can we remove `userInfo` and `errorCode` since we're not using them?

**Review by richherrera — APPROVED** [review]

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift [hunk 4]
```diff
@@ -122,14 +122,20 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
...
 
             if validateResponse.errorNumber == 1050 {
                 errorCode = BTThreeDSecureError.failedAuthentication("").errorCode
+                completionHandler(nil, BTThreeDSecureError.failedAuthentication(""))
+            } else if validateResponse.actionCode == .timeout {
+                errorCode = BTThreeDSecureError.exceededTimeoutLimit.errorCode
+                userInfo = [NSLocalizedDescriptionKey: BTThreeDSecureError.exceededTimeoutLimit.localizedDescription]
+                completionHandler(nil, BTThreeDSecureError.exceededTimeoutLimit)
+            } else {
+                completionHandler(nil, BTThreeDSecureError.exceededTimeoutLimit)
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:136]: I don't think we want our catch all else block to throw the timeout error

Comment (jaxdesmarais) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:136]: I think also we can on line 124 break out the `case .error, .timeout:` into 2 separate cases where the new timeout case can throw the new error you added.

Comment (stechiu) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:136]: That's a good idea, makes it cleaner

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift (hunk 3 above)
Comment (stechiu) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:125]: `userInfo` and `errorCode` get used within the case so we'll need to keep them... I can move `errorCode` as a global variable. Since I'm splitting off `.error` and `.timeout`, I'll keep `userInfo` in as local variable

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift [hunk 5]
```diff
@@ -121,15 +122,22 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
...
             }
             apiClient.sendAnalyticsEvent(BTThreeDSecureAnalytics.challengeFailed)
-            completionHandler(nil, NSError(domain: BTThreeDSecureError.errorDomain, code: errorCode, userInfo: userInfo))
+        case .timeout:
+            let timeoutUserInfo = [NSLocalizedDescriptionKey: BTThreeDSecureError.exceededTimeoutLimit.localizedDescription]
+
+            if validateResponse.actionCode == .timeout {
+                errorCode = BTThreeDSecureError.exceededTimeoutLimit.errorCode
+                completionHandler(nil, NSError(domain: timeoutUserInfo.description, code: errorCode, userInfo: timeoutUserInfo)
+                )
```
Comment (richherrera) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:139]:
```suggestion
                completionHandler(nil, NSError(domain: timeoutUserInfo.description, code: errorCode, userInfo: timeoutUserInfo))
```

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift [hunk 6]
```diff
@@ -122,14 +122,20 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
...
+            var userInfo = [NSLocalizedDescriptionKey: validateResponse.errorDescription]
             var errorCode: Int = BTThreeDSecureError.unknown.errorCode
 
             if validateResponse.errorNumber == 1050 {
                 errorCode = BTThreeDSecureError.failedAuthentication("").errorCode
+                completionHandler(nil, BTThreeDSecureError.failedAuthentication(""))
+            } else if validateResponse.actionCode == .timeout {
+                errorCode = BTThreeDSecureError.exceededTimeoutLimit.errorCode
+                userInfo = [NSLocalizedDescriptionKey: BTThreeDSecureError.exceededTimeoutLimit.localizedDescription]
+                completionHandler(nil, BTThreeDSecureError.exceededTimeoutLimit)
```
[the remark this answers is not in the document]
Comment (stechiu) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift]: I'll create a ticket to add tests for this and `BTThreeDSecureV2Provider`. There doesn't seem to be a straightforward way to test cardinal responses or action codes

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift (hunk 5 above)
Comment (stechiu) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:139]: thanks!

**Review by sshropshire — APPROVED** [review]
👏

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,6 +6,7 @@
 * BraintreeCore
   * Update `URLOpener.openURL(_:completionHandler:)` protocol method to fix method signature change in Xcode 16 beta 3 (fixes #1359)
 * BraintreeThreeDSecure
+  * Add error code and error message for `exceededTimeoutLimit`
```
Comment (jaxdesmarais) [inline CHANGELOG.md:9]: This needs to be moved to `unreleased`

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift [hunk 8]
```diff
@@ -121,15 +122,21 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
...
+                completionHandler(nil, NSError(domain: BTThreeDSecureError.errorDomain, code: errorCode, userInfo: errorUserInfo))
             }
             apiClient.sendAnalyticsEvent(BTThreeDSecureAnalytics.challengeFailed)
-            completionHandler(nil, NSError(domain: BTThreeDSecureError.errorDomain, code: errorCode, userInfo: userInfo))
+        case .timeout:
+            let timeoutUserInfo = [NSLocalizedDescriptionKey: BTThreeDSecureError.exceededTimeoutLimit.localizedDescription]
+
+            if validateResponse.actionCode == .timeout {
+                errorCode = BTThreeDSecureError.exceededTimeoutLimit.errorCode
+                completionHandler(nil, NSError(domain: timeoutUserInfo.description, code: errorCode, userInfo: timeoutUserInfo))
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:138]:
```suggestion
            completionHandler(nil, BTThreeDSecureError.exceededTimeoutLimit)
```

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift [hunk 9]
```diff
@@ -14,6 +14,7 @@ class BTThreeDSecureV2Provider {
 
     var lookupResult: BTThreeDSecureResult? = nil
     var completionHandler: (BTThreeDSecureResult?, Error?) -> Void = { _, _ in }
+    var errorCode: Int = BTThreeDSecureError.unknown.errorCode
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:17]: I think we can revert this change

**Review by jaxdesmarais — CHANGES_REQUESTED** [review]
Blocking for now since there are some changes that would only return an error to the merchant for 1 error number. This would cause issues for merchants getting other errors.

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift [hunk 10]
```diff
@@ -121,15 +122,21 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
                 forResult: lookupResult,
                 completion: completionHandler
             )
-        case .error, .timeout:
-            let userInfo = [NSLocalizedDescriptionKey: validateResponse.errorDescription]
-            var errorCode: Int = BTThreeDSecureError.unknown.errorCode
+        case .error:
+            let errorUserInfo = [NSLocalizedDescriptionKey: validateResponse.errorDescription]
```
[the remark this answers is not in the document]
Comment (stechiu) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:126]: I just realized I left out this `else` statement that was part of the original `.error` case

```
else {
   completionHandler(nil, NSError(domain: BTThreeDSecureError.errorDomain, code: errorCode, userInfo: userInfo))
```

**Comment by stechiu** [comment]
> Blocking for now since there are some changes that would only return an error to the merchant for 1 error number. This would cause issues for merchants getting other errors.

I've added back this missing `else` statement from the original code to the `.error` case. That will fix the issue
```
else {
   completionHandler(nil, NSError(domain: BTThreeDSecureError.errorDomain, code: errorCode, userInfo: userInfo))
```

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift (hunk 8 above)
Comment (stechiu) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:138]: I tried running your suggestion and setting a breakpoint at `validateResponse.actionCode`. The `errorCode` returns as `0` instead of `9` when I reach the max timeout limited

Comment (stechiu) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:138]: `9` is the error code for `.exceededTimoutLimit` and it's what was missing before. I know that, in theory, the suggested refactored code should work. I'm using `po errorCode` in the debugger but might need to test something else that returns the correct error code and error description, so let me know

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift (hunk 9 above)
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:17]:
```suggestion
```

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift (hunk 10 above)
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:126]:
```suggestion
            var errorCode: Int = BTThreeDSecureError.unknown.errorCode
            let errorUserInfo = [NSLocalizedDescriptionKey: validateResponse.errorDescription]
```

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift [hunk 11]
```diff
@@ -121,15 +122,21 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
...
 
             if validateResponse.errorNumber == 1050 {
                 errorCode = BTThreeDSecureError.failedAuthentication("").errorCode
+                completionHandler(nil, NSError(domain: BTThreeDSecureError.errorDomain, code: errorCode, userInfo: errorUserInfo))
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:130]:
```suggestion
```

File: Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift [hunk 12]
```diff
@@ -121,15 +122,21 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
...
             }
             apiClient.sendAnalyticsEvent(BTThreeDSecureAnalytics.challengeFailed)
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift:132]:
```suggestion
            apiClient.sendAnalyticsEvent(BTThreeDSecureAnalytics.challengeFailed)
            completionHandler(nil, NSError(domain: BTThreeDSecureError.errorDomain, code: errorCode, userInfo: userInfo))

```

**Review by jaxdesmarais — APPROVED** [review]
👏 Thanks for addressing feedback and getting a more specific error message added for our merchants!
